emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
+        (thisVals.toSet diff thatVals.toSet).isEmpty
+    }
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    scala.util.hashing.MurmurHash3.unorderedHash(fetch())
+
+  override def toString: String =
+    fetch().toString
```

Comment:
My thinking here was that we would like to hide it from the user as much as possible that these are distributed collections. Therefore, we probably don't want different semantics for the different backends (e.g. something like `DataSet@45754` in case of Flink).

But I just had a new idea: Some REPL-style environments (e.g. Maple) use the following strategy:
- if the result is small then they fully print it
- otherwise they print only a small part of it, and then print `<and 234627635 more elements>`

Maybe we could also do it like this.